pubsublite/internal/wire/publish_data.go

```diff
@@ -0,0 +1,190 @@
+	msgBundler.DelayThreshold = settings.DelayThreshold
+	msgBundler.BundleCountThreshold = settings.CountThreshold
+	msgBundler.BundleByteThreshold = settings.ByteThreshold       // Soft limit
+	msgBundler.BundleByteLimit = MaxPublishRequestBytes           // Hard limit
+	msgBundler.HandlerLimit = 1                                   // Handle batches serially for ordering
+	msgBundler.BufferedByteLimit = settings.BufferedByteLimit * 2 // Effectively disabled, handled in the batcher
+
+	batcher.msgBundler = msgBundler
+	return batcher
+}
+
+func (b *publishMessageBatcher) AddMessage(msg *pb.PubSubMessage, onResult PublishResultFunc) error {
+	msgSize := proto.Size(msg)
+	switch {
+	case msgSize > MaxPublishMessageBytes:
+		return fmt.Errorf("pubsublite: serialized message size is %d bytes, maximum allowed size is MaxPublishMessageBytes (%d)", msgSize, MaxPublishMessageBytes)
+	case msgSize > b.availableBufferBytes:
+		return ErrOverflow
+	}
+
+	holder := &messageHolder{msg: msg, size: msgSize, onResult: onResult}
+	if err := b.msgBundler.Add(holder, msgSize); err != nil {
+		// As we've already checked the size of the message and overflow, the
+		// bundler should not return an error.
+		return fmt.Errorf("pubsublite: failed to batch message: %v", err)
+	}
+	b.availableBufferBytes -= msgSize
+	return nil
+}
+
+func (b *publishMessageBatcher) AddBatch(batch *publishBatch) {
+	b.publishQueue.PushBack(batch)
+}
+
+func (b *publishMessageBatcher) OnPublishResponse(firstOffset int64) error {
+	frontElem := b.publishQueue.Front()
+	if frontElem == nil {
+		return errPublishQueueEmpty
+	}
+	if firstOffset < b.minExpectedNextOffset {
+		return fmt.Errorf("pubsublite: server returned publish response with inconsistent start offset = %d, expected >= %d", firstOffset, b.minExpectedNextOffset)
+	}
+
+	batch, _ := frontElem.Value.(*publishBatch)
+	for i, msgHolder := range batch.msgHolders {
+		// Messages are ordered, so the offset of each message is firstOffset + i.
+		pm := &PublishMetadata{Partition: b.partition, Offset: firstOffset + int64(i)}
+		msgHolder.onResult(pm, nil)
+		b.availableBufferBytes += msgHolder.size
+	}
+
+	b.minExpectedNextOffset = firstOffset + int64(len(batch.msgHolders))
+	b.publishQueue.Remove(frontElem)
+	return nil
+}
+
+func (b *publishMessageBatcher) OnPermanentError(err error) {
+	for elem := b.publishQueue.Front(); elem != nil; elem = elem.Next() {
+		if batch, ok := elem.Value.(*publishBatch); ok {
+			for _, msgHolder := range batch.msgHolders {
+				msgHolder.onResult(nil, err)
+			}
+		}
+	}
+	b.publishQueue.Init()
+}
+
+func (b *publishMessageBatcher) InFlightBatches() []*publishBatch {
+	var batches []*publishBatch
+	for elem := b.publishQueue.Front(); elem != nil; elem = elem.Next() {
+		if batch, ok := elem.Value.(*publishBatch); ok {
+			batches = append(batches, batch)
+		}
+	}
+	return batches
+}
+
+func (b *publishMessageBatcher) Flush() {
+	b.msgBundler.Flush()
+}
+
+func (b *publishMessageBatcher) Done() bool {
```

Comment:
Don't call this "Done". Thats not what it means. It just means there are no publishes outstanding. Call it "NoPublishesOutstanding" or equivalent

Reply:
Done